### PR TITLE
changed barcsq to a vector to allow each factor to have a different inlier threshold

### PR DESCRIPTION
**Pull request by lucacarlone**
I realized that having a default inlier threshold equal to 1 does not make sense for the gtsam user since:
1) "1" is not a reasonable error threshold (for Gaussian noise, we should use the chi2inv to establish a reasonable maximum error) and 
2) the factors can have different dimensions, hence it is desirable to be able to set a different threshold for each factor.

The PR involves:

- [x] making barcsq into a vector
- [x] adding/adjusting the unit tests accordingly
- [x] initializing barcsq using chi2inv

### CODE REVIEW

**Comment by lucacarlone** [comment]
@dellaert @jingnanshi @pantonante : I completed a small enhancement for GNC. please take a look and approve if you like it.

@dellaert : please take a look at the 2 lines commented at line 658 of testGncOptimizer. It seems that "dim()" segfaults when we use the default noise model in the factor. This might be a bug in gtsam.

**Review by dellaert — APPROVED** [review]
Nice! Sorry about delay, teaching large class ;-)

**Review by dellaert — CHANGES_REQUESTED** [review]
PR is fine, but changing to "request changes" as I just noticed CI fails on Linux. Probably something small but needs to be fixed before we can merge.

**Comment by dellaert** [comment]
> @dellaert : please take a look at the 2 lines commented at line 658 of testGncOptimizer. It seems that "dim()" segfaults when we use the default noise model in the factor. This might be a bug in gtsam.

Is that why CI fails? Let me check.

**Comment by lucacarlone** [comment]
@dellaert: let me know if you get some insight into what's failing in CI. I took another pass and cleaned up the comments. @jingnanshi  @pantonante : feel free to approve if this looks good.

File: gtsam/nonlinear/GncOptimizer.h
Comment (jingnanshi) [inline gtsam/nonlinear/GncOptimizer.h:32]: We might want to remove this line.

**Review by jingnanshi — APPROVED** [review]
Everything looks good! I left one comment regarding a commented out include file.

File: gtsam/nonlinear/GncOptimizer.h
Comment (lucacarlone) [reply gtsam/nonlinear/GncOptimizer.h]: good catch - done!

**Comment by dellaert** [comment]
Still some failures on Ubuntu, though?

**Comment by lucacarlone** [comment]
@dellaert : I only changed comments so far, I hoped you could look into the dim() issue. Let me know if that's not feasible.

**Comment by dellaert** [comment]
> @dellaert : I only changed comments so far, I hoped you could look into the dim() issue. Let me know if that's not feasible.

Dim issue is separate, I think. CI failure is unrelated

**Comment by lucacarlone** [comment]
@dellaert : @jingnanshi fixed the issue! at this point, this branch should be good to go

**Review by dellaert — APPROVED** [review]

**Pull request merged**
